### ISSUE TITLE
Titlebar widgets are not garbage collected properly

### ISSUE THREAD

**Issue by sigprof** [awesomeWM/awesome#2830]
Output of `awesome --version`:

```
awesome v4.3-314-g1116bc47 (Too long)
 • Compiled against Lua 5.3.4 (running with Lua 5.3)
 • D-Bus support: ✔
 • xcb-errors support: ✘
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2
```

**How to reproduce the issue:**

Run `awesome` with this patch to the default configuration:

```diff
diff --git a/awesomerc.lua b/awesomerc.lua
index 5d2bd2c1..f9fa69f9 100644
--- a/awesomerc.lua
+++ b/awesomerc.lua
@@ -344,6 +344,9 @@ globalkeys = gears.table.join(
 
 -- @DOC_CLIENT_KEYBINDINGS@
 clientkeys = gears.table.join(
+    awful.key({ modkey, "Shift" }, "t", function(c)
+        c:emit_signal("request::titlebars", "rc", {})
+    end),
     awful.key({ modkey,           }, "f",
         function (c)
             c.fullscreen = not c.fullscreen
```

Open a window (e.g., terminal) and use the `modkey+Shift+t` binding on it.

**Actual result:**

When `modkey+Shift+t` is pressed, the `awesome` process memory consumption continuously increases (with the key repeat rate of 25 the speed was about 10MB per second).

~~I tried to force Lua GC twice and check the Lua memory consumption:~~

~~`awesome-client 'return collectgarbage() and collectgarbage() and collectgarbage("count")'`~~

~~The result was about 11800 and did not significantly increase after the test, and after pressing `modkey+Shift+t` again the process memory usage started to grow immediately (so the Lua GC did not actually free any resources).~~

Sorry, that part of test was wrong — `awesome-client` was connecting to a wrong instance of `awesome` 🙈

I also tested `awful.titlebar.toggle`, and it does not seem to leak.

**Expected result:**

Reconfiguring existing client titlebars should be possible without leaking memory.

**Comment by sigprof** [comment]
Sorry, I messed up my testing, and this seems to be a Lua-level resource leak due to this kind of code in `awful.titlebar`:

```lua
function titlebar.widget.titlewidget(c)
    local ret = textbox()
    local function update()
        ret:set_text(c.name or titlebar.fallback_name)
    end
    c:connect_signal("property::name", update)
    update()

    return ret
end
```

(and an even more complex `update` function in `titlebar.widget.button`).

When this code is executed repeatedly, it creates an unlimited number of signal connections which are never disconnected, and the closures keep both the widget and the client object alive; all those resources are released only when the client gets closed.

Should this problem be fixed in `awful.titlebar` somehow (e.g., by doing tricks with weak tables, like in `awful.widget.taglist`)?

**Comment by sigprof** [comment]
Unfortunately, `client` objects do not support `weak_connect_signal`, otherwise the problem could be fixed very easily (assuming that Lua GC can handle a reference loop between a table and a closure stored in that table), at least for the widget reference part (the client object would still be kept alive by existing widgets, but the `unmanage` signal handler should clean out references to those widgets kept by the `awful.titlebar` code).

Looking at #520, I wonder what *exactly* is the problem with Lua 5.1 GC — is it confined just to functions, or would it also manifest in some form if I store a reference to a table (widget) in a weak table? The code in other places (`awful.widget.clienticon`, `awful.widget.taglist`) stores references to widget instances in weak tables without any strange workarounds.

**Comment by psychon** [comment]
[I just got back from vacation so just a quick answer to the question without me having taken a close look at anything.)

> Looking at #520, I wonder what exactly is the problem with Lua 5.1 GC — is it confined just to functions, or would it also manifest in some form if I store a reference to a table (widget) in a weak table?

IIRC, the problem was in the order in which things happen. In Lua 5.1:
- Object X becomes garbage
- Object X is finalised ("its destructor runs")
- Object X is still garbage and is removed from all weak tables.

In other Lua versions, the object is removed from all weak tables before it is finalised (again: IIRC).
And to make things a bit more complicated: Userdata is removed before its finaliser runs.

The effect of this is that (in Lua 5.1), a drawin could be finalised and afterwards it is still reachable via weak tables and thus we could try redrawing a drawin that was already destroyed.

The short version: Since this issue is about updating a textbox, I do not think that the problem from #520 applies here.

**Comment by psychon** [comment]
> +    awful.key({ modkey, "Shift" }, "t", function(c)
> +        c:emit_signal("request::titlebars", "rc", {})
> +    end),

Not really directly related, but I would just add a titlebar to every client in the `manage` signal (and then perhaps `awful.titlebar.hide` it, if necessary) and then connect a key to `awful.titlebar.toggle`. It does not change anything about the underlying issue here, but it would at least make things work for you (and perhaps it is also more elegant to hide and show a titlebar, instead of hide and recreate-and-destroy-the-old-one).

**Comment by sigprof** [comment]
That keybinding was made only for testing; the actual code was changing the layout of titlebars (all 4 of them) based on many conditions, and just recreating all titlebars with `bgimage` functions capturing sizes of frame elements stored in local variables seemed like a simplest way to write this code.

I did not check whether it is possible to change the size of a titlebar while keeping the assigned widgets; if this works, I could try to change my code so that it does not recreate widgets (e.g., by storing parameters in some table which could be updated without recreating all `bgimage` functions).

**Comment by psychon** [comment]
> That keybinding was made only for testing; 

Ah, okay. Sorry for the useless suggestion then.

> I did not check whether it is possible to change the size of a titlebar 

Not easily. The un-easy (well, unintuitive and likely undocumented) way is something like `c:titlebar_top(20)` to resize the top titlebar to size 20. (Same for `titlebar_right`, `titlebar_left`, and `titlebar_bottom`)
